### PR TITLE
Remove deprecated code before PCL 1.13.0 release

### LINKED ISSUE #5313: Remove deprecated stuff before PCL 1.13.0 release
```
cmake/pcl_find_vtk.cmake:					  "Support of the deprecated backend will be dropped with PCL 1.13.")
common/include/pcl/make_shared.h:PCL_DEPRECATED_HEADER(1, 13, "Use <pcl/memory.h> instead.")
common/include/pcl/point_traits.h:PCL_DEPRECATED_HEADER(1, 13, "Use <pcl/type_traits.h> instead.")
filters/include/pcl/filters/passthrough.h:      PCL_DEPRECATED(1, 13, "use inherited FilterIndices::setNegative() instead")
filters/include/pcl/filters/passthrough.h:      PCL_DEPRECATED(1, 13, "use inherited FilterIndices::getNegative() instead")
filters/include/pcl/filters/impl/crop_hull.hpp:PCL_DEPRECATED(1, 13, "This is a trivial call to base class method")
registration/include/pcl/registration/bfgs.h:  PCL_DEPRECATED(1, 13, "Use `testGradient()` instead")
visualization/include/pcl/visualization/vtk/vtkVertexBufferObjectMapper.h:class PCL_DEPRECATED(1, 13, "The OpenGL backend of VTK is deprecated. Please switch to the OpenGL2 backend.")
visualization/include/pcl/visualization/vtk/vtkVertexBufferObject.h:class PCL_DEPRECATED(1, 13, "The OpenGL backend of VTK is deprecated. Please switch to the OpenGL2 backend.")

```

### PR DESCRIPTION
**Pull request by theoniko**
Fixes #5313

### CODE REVIEW

**Comment by larshg** [comment]
Seems we need to upgrade 18.04 to use VTK 7 first. https://github.com/PointCloudLibrary/pcl/blob/a92da8b793cda3567b89ce86b2f9ed9215e1918d/.ci/azure-pipelines/env.yml#L47

Should be available: https://packages.ubuntu.com/bionic/libvtk7-dev

**Comment by mvieth** [comment]
To be honest I don't really understand why OpenGL compatibility should be deprecated/removed.
PR https://github.com/PointCloudLibrary/pcl/pull/4065 by @SunBlack introduced the deprecation, claiming that we don't loose any compatibility by dropping support of the old backend, and that Ubuntu 18.04 already has OpenGL2 in its repo. However VTK 6.3 in Ubuntu 18.04 and 20.04 both still use OpenGL. So if we remove support for OpenGL, we effectively remove support for VTK 6.3 in Ubuntu, for which I don't see a reason. When a user installs VTK from a package manager (not self compiled), they don't really have a choice which backend they get. And even if the OpenGL backend is not supported in newer VTK versions, I don't see why we have to do anything on the PCL side: for people using VTK from package managers, they will automatically get OpenGL2 for newer VTK versions, and for people self-compiling VTK, they will get warnings or won't be able to build VTK if they try to use OpenGL with newer VTK versions.
So I would propose to un-deprecate the OpenGL backend and continue supporting it. Thoughts @larshg @SunBlack ?

**Comment by mvieth** [comment]
@theoniko Oh and there are some more deprecated things in filters and registration (see the issue). Would you also remove those, please?

**Comment by SunBlack** [comment]
The idea behind dropping the old backend is to simplify the code by removing `#if VTK_RENDERING_BACKEND_OPENGL_VERSION < 2`, as I don't think any maintainer here really checks every relevant change regarding VTK to see if it still works with the old backend.

But in general I think: If you absolutely want to upgrade to PCL 1.13.0, you should also be able to switch to e.g. VTK 7 or in case of doubt build VTK yourself, if you insist on historical in combinations with current systems ;-)

**Comment by mvieth** [comment]
Ok, that sounds reasonable. Then we have to update the 18.04 docker image to VTK 7, as Lars suggested, but in a separate pull request. @theoniko would you also be interested to do that? And please remove the deprecated function in bfgs.h

File: doc/tutorials/content/sources/passthrough/passthrough.cpp
Comment (mvieth) [inline doc/tutorials/content/sources/passthrough/passthrough.cpp:34]: Please don't simply remove this, instead replace with `setNegative`

Comment (theoniko) [reply doc/tutorials/content/sources/passthrough/passthrough.cpp:34]: I thought it was an outdated comment from the context and it did not matter if removed or replaced. Could you explain why it is needed in general because i still cannot make the connection?

**Comment by theoniko** [comment]
@mvieth: PR is not yet ready anyway. I can give it a try to update 18.04 docker image to VTK 7 with a new pr.

File: doc/tutorials/content/sources/passthrough/passthrough.cpp
Comment (mvieth) [reply doc/tutorials/content/sources/passthrough/passthrough.cpp:34]: See [this tutorial](https://pcl.readthedocs.io/projects/tutorials/en/master/passthrough.html) at the very bottom. Also, adding or removing lines in tutorial code is a bit tricky because the tutorials refer to the code by line numbers in the .rst files.
If you want to signal that a PR is not yet ready and we should wait with the reviews, you can mark it as a draft PR.

**Comment by SunBlack** [comment]
@theoniko : Can you also removing all checks for `#if VTK_RENDERING_BACKEND_OPENGL_VERSION < 2` as you removed the sipport for this in CMake as well?

**Comment by mvieth** [comment]
@theoniko Are you still working on this?

**Comment by mvieth** [comment]
> @mvieth: I am quite busy these days. I will try to have a look at the weekend. Is there a fixed date plan for release of 1.13.0

Not a totally fixed date, but somewhat soon, the PCL 1.12.0 release was over a year ago. And we will definitely have to remove the deprecated stuff before the release. If you currently don't have much time, it's also fine if you remove everything but the stuff related to VTK and the VTK rendering backend (and revert the changes in pcl_find_vtk.cmake). As soon as this PR is in a good state, we can merge it, and someone else can remove the remaining things.

File: filters/include/pcl/filters/crop_hull.h
Comment (SunBlack) [inline filters/include/pcl/filters/crop_hull.h:138]: This is a breaking change, as someone external could inherit this class and override it.

File: .ci/azure-pipelines/env.yml
Comment (mvieth) [inline .ci/azure-pipelines/env.yml:47]: This has to be done in a separate pull request, otherwise the docker images won't be updated and some checks in this PR will continue to fail

File: filters/include/pcl/filters/crop_hull.h
Comment (theoniko) [reply filters/include/pcl/filters/crop_hull.h:138]: Ok, I will revert it.

File: .ci/azure-pipelines/env.yml
Comment (theoniko) [reply .ci/azure-pipelines/env.yml:47]: https://github.com/PointCloudLibrary/pcl/pull/5440

**Comment by theoniko** [comment]
@mvieth , @SunBlack [ In Cmakelist under visualization folder ](https://github.com/PointCloudLibrary/pcl/blob/master/visualization/CMakeLists.txt#L231) do i need to update anything to this line? At first look, i do not think so but could you confirm

**Comment by mvieth** [comment]
> @mvieth , @SunBlack [ In Cmakelist under visualization folder ](https://github.com/PointCloudLibrary/pcl/blob/master/visualization/CMakeLists.txt#L231) do i need to update anything to this line? At first look, i do not think so but could you confirm

No, I would also say that this can stay as it is

File: .ci/azure-pipelines/env.yml
Comment (mvieth) [inline .ci/azure-pipelines/env.yml:47]: Please don't revert your earlier PR, leave it at `7`

File: .ci/azure-pipelines/env.yml
Comment (SunBlack) [reply .ci/azure-pipelines/env.yml]: > > @mvieth , @SunBlack [ In Cmakelist under visualization folder ](https://github.com/PointCloudLibrary/pcl/blob/master/visualization/CMakeLists.txt#L231) do i need to update anything to this line? At first look, i do not think so but could you confirm
> 
> No, I would also say that this can stay as it is

Same here. The code looks like it is generally needed for VTK6 under 18.04 and has nothing to do with the legacy backend. You could test what happens if you remove it, but I think you can leave it as it is, since probably the support of VTK6 will be gone with the end of Ubuntu 18.04 next year anyway.

**Comment by SunBlack** [comment]
> > @mvieth , @SunBlack [ In Cmakelist under visualization folder ](https://github.com/PointCloudLibrary/pcl/blob/master/visualization/CMakeLists.txt#L231) do i need to update anything to this line? At first look, i do not think so but could you confirm
> 
> No, I would also say that this can stay as it is

Same here. The code looks like it is generally needed for VTK6 under 18.04 and has nothing to do with the legacy backend. You could test what happens if you remove it, but I think you can leave it as it is, since probably the support of VTK6 will be gone with the end of Ubuntu 18.04 next year anyway.

File: visualization/src/pcl_visualizer.cpp
Comment (mvieth) [inline visualization/src/pcl_visualizer.cpp:1210]: Seems like this line with the `}` has to be removed

Comment (theoniko) [reply visualization/src/pcl_visualizer.cpp]: Done

File: .ci/azure-pipelines/env.yml
Comment (theoniko) [reply .ci/azure-pipelines/env.yml]: Done

**Comment by mvieth** [comment]
Please also remove the deprecated function in bfgs.h

**Review by mvieth — DISMISSED** [review]
Looks good to me now. Thank you!

@larshg and @SunBlack Are you also happy with the changes?

File: cmake/pcl_find_vtk.cmake
Comment (larshg) [inline cmake/pcl_find_vtk.cmake:79]:
```suggestion
if(VTK_VERSION VERSION_GREATER_EQUAL 7.0))
```

**Comment by larshg** [comment]
Only a minor change, else looks good 👍

File: cmake/pcl_find_vtk.cmake
Comment (SunBlack) [reply cmake/pcl_find_vtk.cmake:79]: In case of VTK 6.2/6.3 these Variables are now undefined. In case my research in #4065 was correct, you could define `OpenGL2` already for the older releases, so you can remove the whole `if`

Comment (theoniko) [reply cmake/pcl_find_vtk.cmake]: ok, i removed the whole if.

**Review by larshg — APPROVED** [review]

**Review by mvieth — APPROVED** [review]

**Comment by larshg** [comment]
Thanks for the contribution! 👍

**Pull request merged**
